crates/bevy_animation/src/animatable.rs

```diff
@@ -16,6 +16,10 @@ pub struct BlendInput<T> {
 }
 
 /// An animatable value type.
+#[diagnostic::on_unimplemented(
```

Comment:
Adding the diagnostic for "simple" cases like this is essentially an assertion that Rust is not capable of producing solid error messages for the straightforward `implement MyTrait for MyStruct` cases. If we implement this for cases like `Animatable`, we are essentially saying "every trait in Bevy should have a custom diagnostic" and in a wider scope "every Rust library should implement custom diagnostics for every trait". That is a _very_ strong statement. If we agree with that, we should start knocking on the Rust team's door immediately to work out how to resolve that problem.

Lets compare the error messages:

### With Diagnostic

![image](https://github.com/bevyengine/bevy/assets/2694663/f71f827c-1633-4b16-8e22-228c461366b5)

### Without Diagnostic

![image](https://github.com/bevyengine/bevy/assets/2694663/1bd00dcd-84be-4fee-b5c8-683d31a94eff)

### Comparison

"With Diagnostic" does read as plain English: 

> `Thing` can not be animated
> invalid animateable

Lets compare to "Without Diagnostic"

> the trait bound `Thing: bevy::prelude::Animatable` is not satisfied
> the trait `bevy::prelude::Animatable` is not implemented for `Thing`

On the surface, "plain english" might feel more user friendly. However I will assert that in practice, this _is not_ more user friendly. "cannot be animated' introduces the new verb "animated" which has no direct technical ties to the code at hand. First: "animation" might happen in other traits. Maybe `Thing` _can be animated_ in other contexts. Additionally, `Animatable` has other functionality outside of "animation", namely `Animatable::post_process`. By using "cannot be animated", we are applying a lossy mental model to a context that might only be doing post processing. "Likewise "invalid animatable", while being the same word as `Animatable`, is not the actual symbol. The user has to "couple" these concepts in their head, instead of directly thinking about `Animatable`.

In both cases we have decoupled ourselves from the actual reality of the code. The "friendly english" obfuscates what the actual scope of the problem is.

"Without diagnostic", on the other hand, tells us directly exactly what is wrong _with the code_ and fully encapsulates the problem space.

I think we should remove all instances of things in this class, and use `on_unimplemented` only in specific cases:

1. Cases where a trait is derivable and that derive is used in a high percentage of cases. 
2. Cases where we have complicated implementations of traits that result in nasty / hard to understand error messages.

Reply:
Yep, I'm on board with that decision.

Reply:
Agreed, this is the main reason I haven't taken this off draft just yet. I started by adding basic implementations to all public traits as a maximal goal to then pair back based on actual utility.

I'm going to remove a large number of these custom messages roughly in line with how you've described (derive traits, particularly ugly error messages, etc.)

The other thing I want to do is contrast this with `bevycheck` and the default error messages to ensure this is a strict improvement over the generic case.